jsonpb/jsonpb.go

```diff
@@ -644,7 +649,105 @@ func (u *Unmarshaler) UnmarshalNext(dec *json.Decoder, pb proto.Message) error {
 // permutations of the related Marshaler.
 func (u *Unmarshaler) Unmarshal(r io.Reader, pb proto.Message) error {
 	dec := json.NewDecoder(r)
-	return u.UnmarshalNext(dec, pb)
+	if err := u.UnmarshalNext(dec, pb); err != nil {
+		return err
+	}
+
+	return checkRequiredFields(pb)
```

Comment:
Would it be possible to work the code into UnmarshalNext, rather than iterate over all the fields a second time? (albeit, yes, fairly fast code.)

Reply:
Possibly.  The current unmarshaling logic does rely on json.Unmarshal to populate fields.  To embed the check for required in there means that it will still need to check for all fields that have been unmarshaled.

But you are correct that doing the check w/in the unmarshaling code may be more efficient.  The unmarshaling code is quite complicated though that I prefer to avoid mixing this logic in right now.  A refactor/rewrite of the unmarshaling code should account for this logic.

Reply:
Even though `UnmarshalNext` is used by almost nobody, why is this check in `Unmarshal` as opposed to `UnmarshalNext`?

Reply:
Oops, thanks for catching.  Fixed.